src/serde_utils.rs

```diff
@@ -0,0 +1,92 @@
+use std::fmt;
+use std::marker::PhantomData;
+use serde::de::{Error, Visitor, Deserialize, SeqAccess};
+use serde::ser::{Serializer, SerializeSeq, Serialize};
+
+/// Map to serializeable representation
+pub trait IntoSerializable {
+    type Output;
+    fn into_serializable(self) -> Self::Output;
+}
+
+/// Map from deserialized representation
+pub trait FromDeserialized : Sized {
+    type Input;
+    fn from_deserialized<E>(input: Self::Input) -> Result<Self, E> where E: Error;
+}
+
+
+
+/// Serde combinator. A sequence visitor that maps deserialized elements
+/// lazily; the visitor can also emit new errors if the elements have errors.
+pub struct MappedSequenceVisitor<T, R, F>
+    where F: Fn(T) -> Result<R, &'static str>
+{
+    f: F,
+    marker: PhantomData<fn() -> T>
```

Comment:
Is this correct? There are no methods that return T.

Reply:
It is used in the Fn to T sense, I don't think it matters that there is no method with the same signature.

It is also used in the Fn Of T to R sense, so the result must be that it is invariant in T. It is not mandatory to specify this exactly for any reason, so I'm content with the approximation here. It does manage to avoid the implications of phantom data of T, at least.